packages/front-end/components/Settings/InviteModal.tsx

```diff
@@ -44,59 +55,144 @@ const InviteModal: FC<{ mutate: () => void; close: () => void }> = ({
   }
 
   const onSubmit = form.handleSubmit(async (value) => {
-    const resp = await apiCall<{
-      emailSent: boolean;
-      inviteUrl: string;
-      status: number;
-      message?: string;
-    }>(`/invite`, {
-      method: "POST",
-      body: JSON.stringify(value),
-    });
+    const inviteArr = value.email.split(",");
 
-    if (resp.emailSent) {
-      mutate();
-      close();
-    } else {
-      setInviteUrl(resp.inviteUrl);
-      setEmailSent(resp.emailSent);
-      mutate();
+    if (canSubscribe && activeAndInvitedUsers + inviteArr.length > freeSeats) {
+      setShowUpgradeModal(true);
+      return;
     }
 
-    track("Team Member Invited", {
-      emailSent,
-      role: value.role,
-    });
-  });
+    for (const email of inviteArr) {
+      const resp = await apiCall<{
+        emailSent: boolean;
+        inviteUrl: string;
+        status: number;
+        message?: string;
+      }>(`/invite`, {
+        method: "POST",
+        body: JSON.stringify({
+          email: email,
+          role: value.role,
+        }),
+      });
+
+      if (resp.emailSent) {
+        setSuccessfulInvites((successfulInvites) => [
+          ...successfulInvites,
+          {
+            email: email,
+            inviteUrl: resp.inviteUrl,
+            emailSent: resp.emailSent,
+          },
+        ]);
+      } else {
+        setFailedInvites((failedInvites) => [
+          ...failedInvites,
+          {
+            email: email,
+            inviteUrl: resp.inviteUrl,
+            emailSent: resp.emailSent,
+          },
+        ]);
```

Comment:
Instead of calling the setState functions directly in a loop, it's better to build the arrays in a variable and then call setState at the end of the loop.  For example:

```ts
const failed: InviteResult[] = [];
const succeeded: InviteResult[] = [];

for (const email of inviteArr) {
  ...
  const result = {
    email, 
    inviteUrl: resp.inviteUrl,
    emailSent: resp.emailSent
  };
  if (resp.emailSent) {
    succeeded.push(result);
  }  else {
    failed.push(result);
  }
}

setSuccessfulInvites(succeeded);
setFailedInvites(failed);
```

Reply:
Ah - that makes sense. I was struggling with that - at first thinking I could use a callback to force React to set the state. I also tried to use `flushSync()` without any success. Thanks for sharing this pattern.